### PR TITLE
Azure Container Apps - 4/28 release

### PR DESCRIPTION
**Pull request by haroonf**
Adds support for
az containerapp up
az containerapp logs show (Log-stream support)
az containerapp exec (SSH support)
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Container

**Comment by panchagnula** [comment]
@zhoxing-ms please give us your initial review feedback & we will address this on Monday (PST hours) for a 4/26 release. Thank you.

File: src/containerapp/azext_containerapp/.flake8
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/.flake8:4]: Why don't you fix the issue of the code style directly but ignore the flake8 checks?

File: src/containerapp/azext_containerapp/_acr_run_polling.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_acr_run_polling.py:36]: The logic of serializing and calling `LROPoller` is usually implemented in the SDK, why don't you use SDK?

File: src/containerapp/azext_containerapp/_params.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_params.py:110]: > type=str

`str` is the default for `type`, so there is no need to declare it again

File: src/containerapp/azext_containerapp/_utils.py
Comment (zhoxing-ms) [inline src/containerapp/azext_containerapp/_utils.py:27]: > \# logger.warning("Creating '%s' role assignment under scope '%s'", role, scope) 
> \# retry till server replication is done

I'm not sure whether it's reasonable to reimplement the whole `create_service_principal_for_rbac` method just to remove unnecessary warning statements. It has a disadvantage that it will cause the same code duplication and increase the maintenance cost. @jiasli Do you have any suggestions for this?

File: src/containerapp/azext_containerapp/_params.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_params.py:110]: @haroonf can you fix this.

File: src/containerapp/azext_containerapp/.flake8
Comment (panchagnula) [reply src/containerapp/azext_containerapp/.flake8:4]: @haroonf can you remove these changes as well - will re-visit this once Silas is back.

File: src/containerapp/azext_containerapp/_utils.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_utils.py:193]: While this works @StrawnSC to give you some context this issue is that, the call the check list & retrieve the [0] once the GitHub action add is completed. Talking to Sanchit the GitHub actions add Async response returns an operationId & we can poll on the operationId, to validate this is 200 & successful, before we call the list of workflows.

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:2071]: NIT: Can we add comment explaining the need for this - so we don't accidentally delete / remove this?

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:1998]: @StrawnSC add some exceptional handling for if the branch info is incorrect or missing, so a user can run the command with --branch (See Jeff's email with more details as well)

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:2022]: Also add the command to change this later. i.e may be containerapp up with --ingress & --target-port & the update command they can run post app is created.

File: src/containerapp/azext_containerapp/_up_utils.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_up_utils.py:707]: Nit: remove () around the app name.

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:2015]: lets add the quick start image as a constant & reference here so if this changes we can easily update this in the constants.

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:2016]: Good to see some defaults added but can we also make sure the Output response includes the FQDN, currently not seeing it.

File: src/containerapp/azext_containerapp/custom.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/custom.py:1064]: Fix the optional branch value here as well from None to main. GitHub actions add is broken with the same behavior as well. @StrawnSC can the existing tests for GitHub-actions be improved to catch the lack of defaults like for branch?

File: src/containerapp/azext_containerapp/_utils.py
Comment (StrawnSC) [reply src/containerapp/azext_containerapp/_utils.py:27]: To be more specific, I reimplemented the method to eliminate a warning statement that said: 
```
The output includes credentials that you must protect. 
Be sure that you do not include these credentials in your code or check the credentials into your source control.
For more information, see https://aka.ms/azadsp-cli
```

In this case, the warning is incorrect because these service principal credentials are not outputted (they are passed to the containerapp GitHub action API)

Comment (jiasli) [reply src/containerapp/azext_containerapp/_utils.py:27]: I agree with @zhoxing-ms. Vendoring `create_service_principal_for_rbac` is a bad idea. `create_service_principal_for_rbac` is updated frequently due to design change or security requirements. Once vendored, you also need to redo Microsoft Graph migration (https://github.com/Azure/azure-cli/issues/22174). For example, `credential_description` and `password` are removed from `create_application`.

If you want the warning not to be shown, I can add a switch.

File: src/containerapp/azext_containerapp/_utils.py
Comment (jiasli) [inline src/containerapp/azext_containerapp/_utils.py:34]: Importing and using private method is definitely a no-go!

**Review by jiasli — CHANGES_REQUESTED** [review]
Shouldn't copy the code of `create_service_principal_for_rbac`.

File: src/containerapp/azext_containerapp/custom.py
Comment (StrawnSC) [reply src/containerapp/azext_containerapp/custom.py:1064]: I should have called this out in a comment, my bad. The reason the help text says the GitHub action defaults to the main branch, but this parameter is None by default is that the API assumes the branch when it is passed as None. When I tested on two of my own repos, the action is created on the main branch when no `--branch` is supplied

File: src/containerapp/azext_containerapp/custom.py
Comment (haroonf) [reply src/containerapp/azext_containerapp/custom.py:2022]: @panchagnula should this be logger.info or warning?

File: src/containerapp/azext_containerapp/_utils.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_utils.py:41]: @jiasli this is something we need for the initial release of the extension for 'up' support &  we have to copy this over since we want some additional logger.warning we want to add. We will move to the new graph calls with the next release, Could you point us to the new Microsoft.Graph migration changes as well, so we can change it to that & do testing for 'up' command? Thanks!

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py:26]: @haroonf in the next release lets update this test to not include the tag - since we had a bug on this a while back would be good to make that the test case. Will file a bug on you for this - not a blocker. Thanks!

**Review by panchagnula — APPROVED** [review]
LGTM

**Comment by panchagnula** [comment]
> Shouldn't copy the code of `create_service_principal_for_rbac`

@jiasli / @zhoxing-ms we need to copy & maintain this code for "up" scenarios, since we want some specific logging & error handling for up. We don't to modify the generic command code, for our purposes.  We add this mainly because with az containerapp up with GitHub-actions we do the best effort to have all the required non ACA resources ready/set up, so they don't need to run multiple separate commands. Hope this answers why we need this here. Thanks!

File: src/containerapp/azext_containerapp/_utils.py
Comment (panchagnula) [reply src/containerapp/azext_containerapp/_utils.py:41]: I found this https://docs.microsoft.com/en-us/cli/azure/microsoft-graph-migration?tabs=powershell#az-ad-app-create @StrawnSC I will add this to our ADO work item so we understand the exact changes we need to make next to ensure we use Microsft.Graph implementation to add SP.

File: src/containerapp/azext_containerapp/_up_utils.py
Comment (panchagnula) [inline src/containerapp/azext_containerapp/_up_utils.py:344]: Thanks for adding this is useful for investigations when GitHub workflow is failing with SP access permissions error. Please add the same when we are fetching & re-using an existing SP. Thanks!

File: src/containerapp/azext_containerapp/_acr_run_polling.py
Comment (zhoxing-ms) [reply src/containerapp/azext_containerapp/_acr_run_polling.py:36]: OK, please try to use Python SDK instead of manually implementing these codes in the future

**Comment by zhoxing-ms** [comment]
In fact, I personally think the main problem of blocking this PR merge is to copy the code of `create_service_principal_for_rbac`, which is really not a good design. 
@panchagnula Personally, I think maybe we can approve and merge this PR first to meet your release schedule, then please submit a PR for refactoring these codes as soon as possible, what do you think of this plan? @panchagnula @jiasli

File: src/containerapp/azext_containerapp/_utils.py
Comment (jiasli) [reply src/containerapp/azext_containerapp/_utils.py:41]: This extension will break with `azure-cli-core` 2.37.0. Please see the upper bound of `azure-cli-core` to 2.36.0.

**Review by jiasli — APPROVED** [review]
Approve as long as `maxCliCoreVersion` is set to 2.36.0.

**Comment by StrawnSC** [comment]
> Approve as long as `maxCliCoreVersion` is set to 2.36.0.

Done, thank you

**Comment by panchagnula** [comment]
> Approve as long as `maxCliCoreVersion` is set to 2.36.0.

Thank you @jiasli & @zhoxing-ms - can we get this merged & released for 4/29 early hours PST .

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**
